File: server/src/services/payment/payment.service.ts
import { Injectable, BadRequestException, NotFoundException, ForbiddenException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Payment } from 'src/models/payment.entity';
import { Group } from 'src/models/group.entity';
import { User } from 'src/models/user.entity';
import { CreatePaymentDto } from 'src/dto/create-payment.dto';
import { ExpenseParticipant } from 'src/models/expense-participant.entity';
import { Expense } from 'src/models/expense.entity';

@Injectable()
export class PaymentService {
  constructor(
    @InjectRepository(Payment) private paymentRepo: Repository<Payment>,
    @InjectRepository(Group) private groupRepo: Repository<Group>,
    @InjectRepository(User) private userRepo: Repository<User>,
    @InjectRepository(ExpenseParticipant) private expPartRepo: Repository<ExpenseParticipant>,
    @InjectRepository(Expense) private expenseRepo: Repository<Expense>,
  ) {}

  private toCentsFromStringOrNumber(value: string | number): number {
    const v = typeof value === 'string' ? parseFloat(value) : value;
    return Math.round(v * 100);
  }

  private centsToNumber(cents: number): number {
    return +(cents / 100).toFixed(2);
  }

  async createPayment(requestor: { id: number | string }, groupId: string, dto: CreatePaymentDto) {
    // Validate group and membership
    const group = await this.groupRepo.findOne({ where: { id: groupId }, relations: ['members', 'members.user'] });
    if (!group) throw new NotFoundException('Group not found');

    const memberIds = new Set(group.members.map((m) => String(m.user.id)));
    if (!memberIds.has(String(dto.toUserId))) throw new BadRequestException('toUser is not a member of the group');
    if (!memberIds.has(String(requestor.id))) throw new ForbiddenException('From user must be a group member');

    if (Number(dto.amount) <= 0) throw new BadRequestException('Amount must be positive');

    const fromUser = await this.userRepo.findOne({ where: { id: String(requestor.id) } });
    const toUser = await this.userRepo.findOne({ where: { id: dto.toUserId } });
    if (!fromUser || !toUser) throw new NotFoundException('Users not found');

    const p = this.paymentRepo.create({
      group,
      fromUser,
      toUser,
      amount: dto.amount.toFixed(2),
      note: dto.note,
    });

    return this.paymentRepo.save(p);
  }

  async listPayments(groupId: string, limit = 50, offset = 0) {
    return this.paymentRepo.find({
      where: { group: { id: groupId } as any },
      relations: ['fromUser', 'toUser'],
      order: { date: 'DESC' },
      take: limit,
      skip: offset,
    });
  }

  /**
   * computeBalances(requestorId, groupId)
   * returns a map: { otherUserId: number } meaning how much the OTHER user owes REQUESTOR (positive -> they owe requestor)
   */
  async computeBalances(requestorId: string | number, groupId: string) {
    // 1) get all group members
    const group = await this.groupRepo.findOne({ where: { id: groupId }, relations: ['members', 'members.user'] });
    if (!group) throw new NotFoundException('Group not found');

    const members = group.members.map((m) => m.user);
    const otherMemberIds = members.filter((u) => String(u.id) !== String(requestorId)).map((u) => String(u.id));

    // Prepare maps (all cents)
    const mapPaidByReq = new Map<string, number>(); // how much requestor paid for others
    const mapPaidForReq = new Map<string, number>(); // how much others paid for requestor
    const mapPaymentsFrom = new Map<string, number>(); // payments from other -> requestor
    const mapPaymentsTo = new Map<string, number>();   // payments from requestor -> other

    // 1) paid_by_requestor_for_others:
    // sum(ep.share) grouped by ep.userId where ep.expense.payerId = requestor
    const paidByReqRows = await this.expPartRepo
      .createQueryBuilder('ep')
      .select('ep.userId', 'userId')
      .addSelect('SUM(ep.share)', 'amount')
      .innerJoin('ep.expense', 'e')
      .where('e.groupId = :groupId', { groupId })
      .andWhere('e.payerId = :req', { req: String(requestorId) })
      .groupBy('ep.userId')
      .getRawMany();

    for (const r of paidByReqRows) {
      // r.amount is string like '30.00'
      const cents = Math.round(parseFloat(r.amount) * 100);
      mapPaidByReq.set(String(r.userId), cents);
    }

    // 2) paid_by_others_for_requestor:
    // where ep.userId = requestor (i.e., requestor was participant), group by e.payerId
    const paidForReqRows = await this.expPartRepo
      .createQueryBuilder('ep')
      .select('e.payerId', 'payerId')
      .addSelect('SUM(ep.share)', 'amount')
      .innerJoin('ep.expense', 'e')
      .where('e.groupId = :groupId', { groupId })
      .andWhere('ep.userId = :req', { req: String(requestorId) })
      .andWhere('e.payerId != :req', { req: String(requestorId) })
      .groupBy('e.payerId')
      .getRawMany();

    for (const r of paidForReqRows) {
      const cents = Math.round(parseFloat(r.amount) * 100);
      mapPaidForReq.set(String(r.payerId), cents);
    }

    // 3) payments grouped (only those payments involving requestor)
    const paymentRows = await this.paymentRepo
      .createQueryBuilder('p')
      .select('p.fromUserId', 'fromUserId')
      .addSelect('p.toUserId', 'toUserId')
      .addSelect('SUM(p.amount)', 'amount')
      .where('p.groupId = :groupId', { groupId })
      .andWhere('(p.fromUserId = :req OR p.toUserId = :req)', { req: String(requestorId) })
      .groupBy('p.fromUserId')
      .addGroupBy('p.toUserId')
      .getRawMany();

    for (const r of paymentRows) {
      const cents = Math.round(parseFloat(r.amount) * 100);
      const from = String(r.fromUserId);
      const to = String(r.toUserId);
      if (to === String(requestorId) && from !== String(requestorId)) {
        // other -> requestor
        mapPaymentsFrom.set(from, (mapPaymentsFrom.get(from) || 0) + cents);
      } else if (from === String(requestorId) && to !== String(requestorId)) {
        // requestor -> other
        mapPaymentsTo.set(to, (mapPaymentsTo.get(to) || 0) + cents);
      }
    }

    // Now compute final per-user balances (in cents)
    const result: Record<string, number> = {};
    for (const otherId of otherMemberIds) {
      const paidByReq = mapPaidByReq.get(otherId) || 0;
      const paidForReq = mapPaidForReq.get(otherId) || 0;
      const paymentsFromOther = mapPaymentsFrom.get(otherId) || 0;
      const paymentsFromReq = mapPaymentsTo.get(otherId) || 0;

      // amount other owes requestor in cents:
      // (requestor paid for other) - (other paid for requestor) - (other -> requestor payments) + (??)
      // We subtract payments both ways: a payment from requestor to other reduces how much other owes requestor (i.e., makes requestor owe more)
      // Using the formula discussed earlier:
      const cents = paidByReq - paidForReq - paymentsFromOther + paymentsFromReq;

      // Interpretation: positive -> other owes requestor. negative -> requestor owes other.
      result[otherId] = +(cents / 100).toFixed(2);
    }

    // Optionally include own totals:
    // totalOwedToRequestor = sum of positive values
    // totalOwingByRequestor = sum of negative values (in absolute)
    let totalOwedToRequestor = 0;
    let totalOwingByRequestor = 0;
    for (const k of Object.keys(result)) {
      const v = result[k];
      if (v > 0) totalOwedToRequestor += v;
      else totalOwingByRequestor += Math.abs(v);
    }

    return {
      requestorId: String(requestorId),
      groupId,
      balances: result,
      totalOwedToRequestor: +totalOwedToRequestor.toFixed(2),
      totalOwingByRequestor: +totalOwingByRequestor.toFixed(2),
    };
  }
}

File: server-java/moneyjar/moneyjar/src/main/java/com/tarungattu/moneyjar/services/payment/PaymentService.java
package com.tarungattu.moneyjar.services.payment;

import com.tarungattu.moneyjar.dto.CreatePaymentDto;
import com.tarungattu.moneyjar.models.Group;
import com.tarungattu.moneyjar.models.GroupMember;
import com.tarungattu.moneyjar.models.Payment;
import com.tarungattu.moneyjar.models.User;
import com.tarungattu.moneyjar.repository.ExpenseParticipantRepository;
import com.tarungattu.moneyjar.repository.GroupRepository;
import com.tarungattu.moneyjar.repository.PaymentRepository;
import com.tarungattu.moneyjar.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class PaymentService {

    @Autowired
    private PaymentRepository paymentRepository;

    @Autowired
    private GroupRepository groupRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ExpenseParticipantRepository expenseParticipantRepository;

    @Transactional
    public Payment createPayment(UUID requestorId, UUID groupId, CreatePaymentDto dto) {
        if (dto.getAmount().compareTo(BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Amount must be positive");
        }

        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));

        List<UUID> memberIds = group.getMembers().stream()
                .map(m -> m.getUser().getId())
                .collect(Collectors.toList());

        if (!memberIds.contains(dto.getToUserId())) {
            throw new RuntimeException("toUser is not a member of the group");
        }
        if (!memberIds.contains(requestorId)) {
            throw new RuntimeException("From user must be a group member");
        }

        User fromUser = userRepository.findById(requestorId).orElseThrow(() -> new RuntimeException("User not found"));
        User toUser = userRepository.findById(dto.getToUserId()).orElseThrow(() -> new RuntimeException("User not found"));

        Payment payment = new Payment();
        payment.setGroup(group);
        payment.setFromUser(fromUser);
        payment.setToUser(toUser);
        payment.setAmount(dto.getAmount());
        payment.setNote(dto.getNote());

        return paymentRepository.save(payment);
    }

    @Transactional(readOnly = true)
    public List<Payment> listPayments(UUID groupId, int limit, int offset) {
        return paymentRepository.findByGroupIdOrderByDateDesc(groupId)
                .stream().skip(offset).limit(limit).collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Map<String, Object> computeBalances(UUID requestorId, UUID groupId) {
        Group group = groupRepository.findById(groupId)
                .orElseThrow(() -> new RuntimeException("Group not found"));

        List<UUID> otherMemberIds = group.getMembers().stream()
                .map(m -> m.getUser().getId())
                .filter(id -> !id.equals(requestorId))
                .collect(Collectors.toList());

        // Prepare maps (cents)
        Map<UUID, Long> mapPaidByReq = new HashMap<>();
        Map<UUID, Long> mapPaidForReq = new HashMap<>();
        Map<UUID, Long> mapPaymentsFrom = new HashMap<>(); // From other TO req
        Map<UUID, Long> mapPaymentsTo = new HashMap<>();   // From req TO other

        // 1. Paid by requestor for others
        List<Object[]> paidByReqRows = expenseParticipantRepository.sumPaidByPayer(groupId, requestorId);
        for (Object[] row : paidByReqRows) {
            UUID userId = (UUID) row[0];
            BigDecimal share = (BigDecimal) row[1];
            mapPaidByReq.put(userId, share.multiply(new BigDecimal("100")).longValue());
        }

        // 2. Paid for requestor by others
        List<Object[]> paidForReqRows = expenseParticipantRepository.sumPaidForUser(groupId, requestorId);
        for (Object[] row : paidForReqRows) {
            UUID payerId = (UUID) row[0];
            BigDecimal share = (BigDecimal) row[1];
            mapPaidForReq.put(payerId, share.multiply(new BigDecimal("100")).longValue());
        }

        // 3. Payments involving requestor
        List<Object[]> paymentRows = paymentRepository.sumPaymentsInvolvingUser(groupId, requestorId);
        for (Object[] row : paymentRows) {
            UUID fromId = (UUID) row[0];
            UUID toId = (UUID) row[1];
            BigDecimal amount = (BigDecimal) row[2];
            long cents = amount.multiply(new BigDecimal("100")).longValue();

            if (toId.equals(requestorId) && !fromId.equals(requestorId)) {
                mapPaymentsFrom.put(fromId, mapPaymentsFrom.getOrDefault(fromId, 0L) + cents);
            } else if (fromId.equals(requestorId) && !toId.equals(requestorId)) {
                mapPaymentsTo.put(toId, mapPaymentsTo.getOrDefault(toId, 0L) + cents);
            }
        }

        Map<String, BigDecimal> balances = new HashMap<>();
        BigDecimal totalOwedToRequestor = BigDecimal.ZERO;
        BigDecimal totalOwingByRequestor = BigDecimal.ZERO;

        for (UUID otherId : otherMemberIds) {
            long paidByReq = mapPaidByReq.getOrDefault(otherId, 0L);
            long paidForReq = mapPaidForReq.getOrDefault(otherId, 0L);
            long paymentsFromOther = mapPaymentsFrom.getOrDefault(otherId, 0L);
            long paymentsFromReq = mapPaymentsTo.getOrDefault(otherId, 0L);

            // Calculation formula from NestJS:
            // cents = paidByReq - paidForReq - paymentsFromOther + paymentsFromReq;
            long balanceCents = paidByReq - paidForReq - paymentsFromOther + paymentsFromReq;
            
            BigDecimal balance = new BigDecimal(balanceCents).divide(new BigDecimal("100"));
            balances.put(otherId.toString(), balance);

            if (balance.compareTo(BigDecimal.ZERO) > 0) {
                totalOwedToRequestor = totalOwedToRequestor.add(balance);
            } else {
                totalOwingByRequestor = totalOwingByRequestor.add(balance.abs());
            }
        }

        Map<String, Object> result = new HashMap<>();
        result.put("requestorId", requestorId.toString());
        result.put("groupId", groupId.toString());
        result.put("balances", balances);
        result.put("totalOwedToRequestor", totalOwedToRequestor);
        result.put("totalOwingByRequestor", totalOwingByRequestor);

        return result;
    }
}
